### PR TITLE
Adds a noise module and a RMS function for signals

### PR DESCRIPTION
**Pull request by paulcadman**
The RMS function is needed because I'd like the amount of noise to be specified using the signal to noise ratio - which requires the RMS.

The first commit just creates the noise module and moves sample to it as it's not used by main.

I also changed `approx_eq` to take a tolerance parameter. Thankfully the callsites of `prop_assert_approx_eq` do not have to change because macros allow overloading (so the two argument version of the macro keeps the same behaviour).

The third commit actually adds the RMS function. This samples the signal with an arbitrary number of 100 samples. The number of samples should really be a function of the input period but this is good enough for now.

### CODE REVIEW

File: src/noise.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+pub fn rms(s: Signal, period: TimeSecs) -> f64 {
+    // The number of samples should depend on the period.
+    let samples = 100;
+    let rate = FrequencyHz(((samples as f64) / period.0) as u32);
+
+    // The fold function is derived from:
+    //
+    // let a_1, ... a_n be the items of the iterator.
+    //
+    // mean_n = (1/n) * \sum_{i=1}^n a_i
```
Comment (mheiber) [inline src/noise.rs]: What's the relationship between `mean_n` and `rms`? Is it `rms = sqrt(mean_100)` because we are taking 100 samples?

File: src/noise.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    //        = (1/n) * ( (n - 1) * mean_{n - 1} + a_n )
+    //        = mean_{n - 1} + (a_n - mean_{n - 1}) / n
```
Comment (mheiber) [inline src/noise.rs]: Questions related to 0:

- May want to add the base case to the pseudocode: `// mean_0 = 0` or `// mean_1 = 1`
- It looks like the pseudocode uses 1-indexing but the real code uses 0-indexing. I suggest using the same indexing for both the real code and the pseudocode.
- The real code has a `+ 1` that is not in the pseudocode. I suspect this may be related to 0-vs-1-indexing.

File: src/noise.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+/// * `period` - The period (starting at phase=0) over which the computation is made
+///
+pub fn rms(s: Signal, period: TimeSecs) -> f64 {
+    // The number of samples should depend on the period.
+    let samples = 100;
+    let rate = FrequencyHz(((samples as f64) / period.0) as u32);
+
+    // The fold function is derived from:
+    //
+    // let a_1, ... a_n be the items of the iterator.
```
Comment (mheiber) [inline src/noise.rs]: I suggest rewording this comment or tweaking the pseudocode so it matches.

The items in the iterator iuc are of a tuple type (`(usize, f64)`), but in the pseudocode `a_n` is scalar.

File: src/noise.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+///
+/// * `s` - The `Signal` used to compute the RMS
+/// * `period` - The period (starting at phase=0) over which the computation is made
+///
+pub fn rms(s: Signal, period: TimeSecs) -> f64 {
+    // The number of samples should depend on the period.
+    let samples = 100;
+    let rate = FrequencyHz(((samples as f64) / period.0) as u32);
+
+    // The fold function is derived from:
```
Comment (mheiber) [inline src/noise.rs]: I really like this comment that gradually morphs the maths into code.
The volume of GitHub-comments are no reflection on the quality of the code comment.

File: src/noise.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    //
+    // mean_n = (1/n) * \sum_{i=1}^n a_i
+    //        = (1/n) * ( (n - 1) * mean_{n - 1} + a_n )
+    //        = mean_{n - 1} + (a_n - mean_{n - 1}) / n
+    //
+    sample(rate, s)
+        .map(|x| x.powf(2.0))
+        .enumerate()
+        .take(samples)
+        .fold(0.0 as f64, |m, (n, x)| {
```
Comment (mheiber) [inline src/noise.rs]: bikeshedding: why `m`? Is it short for `memo`?

When I was trying to match the code up with the pseudocode, I had to keep a translation table. I take it `m` is `a_{n-1}`?

How about `running_mean`?

File: src/noise.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+pub fn sample(rate: FrequencyHz, s: Signal) -> impl Iterator<Item = f64> {
+    (0..).map(move |n: u32| {
+        let sample_period = rate.period();
+        let t = n as f64 * sample_period.0;
+        s.at(TimeSecs(t))
+    })
+}
+
+/// Estimates the root mean square (RMS) of a signal over a period
```
Comment (mheiber) [inline src/noise.rs:11]: Want to give a reference?
Looks like what we have here matches: https://en.wikipedia.org/wiki/Root_mean_square

File: tests/tests.rs [hunk 7]
```diff
@@ -95,5 +127,12 @@ mod tests {
                 prop_assert_approx_eq!(v, expected);
             }
         }
+
+        #[test]
+        fn rms_square_wave(f in arb_frequency(), p in arb_timesecs()) {
+            prop_assume!(f > FrequencyHz(0));
+            prop_assume!(p > TimeSecs(0.0));
+            prop_assert_approx_eq!(rms(square_wave(f), p), 1.0);
```
Comment (mheiber) [inline tests/tests.rs:135]: I tried translating the pseudocode to Rust naively, thinking it could be useful as a test oracle. Oddly, the tests fail when I replace `rms` with the following:

```rust
/// Estimates the root mean square (RMS) of a signal over a period
///
/// # Arguments
///
/// * `s` - The `Signal` used to compute the RMS
/// * `period` - The period (starting at phase=0) over which the computation is made
///
pub fn rms(s: Signal, period: TimeSecs) -> f64 {
    // The number of samples should depend on the period.
    let samples = 100;
    let rate = FrequencyHz(((samples as f64) / period.0) as u32);
    let a: Vec<_> = sample(rate, s).map(|x| x.powf(2.0)).take(samples).collect();

    fn mean(a: Vec<f64>, n: usize) -> f64 {
        if n == 0 {
            0.0
        }
        else {
            let a_n = a[n];
            let mean_n_minus_1 = mean(a, n - 1);
            mean_n_minus_1 + (a_n - mean_n_minus_1) / ((n as i32) as f64)
        }
    }
    let n = a.len() - 1;
    mean(a, n).sqrt()
}
```

Just an observation, not really actionable.

**Review by mheiber — APPROVED** [review]

**Comment by mheiber** [comment]
Nice

File: src/noise.rs (hunk 3 above)
Comment (paulcadman) [reply src/noise.rs]: In the end I think the mix of maths and pseudo-code is too confusion so I decided to remove it.

File: src/noise.rs (hunk 5 above)
Comment (paulcadman) [reply src/noise.rs]: `m` is indeed the running mean. `x` is `a_n` and `n` is `n`.

File: src/noise.rs (hunk 1 above)
Comment (paulcadman) [reply src/noise.rs]: Yes - the pseudo-code only described the `fold` closure.

File: tests/tests.rs (hunk 7 above)
Comment (paulcadman) [reply tests/tests.rs:135]: I had a look but I can't see what's going wrong here.
